apps/emqx_resource/src/emqx_resource_manager.erl

```diff
@@ -0,0 +1,427 @@
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+-module(emqx_resource_manager).
+-behaviour(gen_statem).
+
+-include("emqx_resource.hrl").
+-include("emqx_resource_utils.hrl").
+
+% API
+-export([
+    ensure_resource/5,
+    create_dry_run/2,
+    ets_lookup/1,
+    get_metrics/1,
+    health_check/1,
+    list_all/0,
+    list_group/1,
+    lookup/1,
+    recreate/4,
+    remove/1,
+    reset_metrics/1,
+    restart/2,
+    set_resource_status_connecting/1,
+    stop/1
+]).
+
+% Server
+-export([start_link/5]).
+
+% Behaviour
+-export([init/1, callback_mode/0, handle_event/4, terminate/3]).
+
+% State record
+-record(data, {id, group, mod, config, opts, status, state, error}).
+
+-define(SHORT_HEALTHCHECK_INTERVAL, 1000).
+-define(HEALTHCHECK_INTERVAL, 15000).
+-define(ETS_TABLE, emqx_resource_manager).
+-define(TIME_DIVISOR, 100).
+-define(WAIT_FOR_RESOURCE_DELAY, 100).
+
+%%------------------------------------------------------------------------------
+%% API
+%%------------------------------------------------------------------------------
+
+%% @doc Called from emqx_resource when starting a resource instance.
+%%
+%% Triggers the emqx_resource_manager_sup supervisor to actually create
+%% and link the process itself if not already started.
+-spec ensure_resource(
+    instance_id(),
+    resource_group(),
+    resource_type(),
+    resource_config(),
+    create_opts()
+) -> {ok, resource_data()}.
+ensure_resource(InstId, Group, ResourceType, Config, Opts) ->
+    case lookup(InstId) of
+        {ok, _Group, Data} ->
+            {ok, Data};
+        {error, not_found} ->
+            do_start(InstId, Group, ResourceType, Config, Opts),
+            {ok, _Group, Data} = lookup(InstId),
+            {ok, Data}
+    end.
+
+%% @doc Called from `emqx_resource` when doing a dry run for creating a resource instance.
+%%
+%% Triggers the `emqx_resource_manager_sup` supervisor to actually create
+%% and link the process itself if not already started, and then immedately stops.
+-spec create_dry_run(resource_type(), resource_config()) ->
+    ok | {error, Reason :: term()}.
+create_dry_run(ResourceType, Config) ->
+    InstId = make_test_id(),
+    _ = emqx_resource_manager_sup:start_child(InstId, <<"dry_run">>, ResourceType, Config, #{}),
+    case wait_for_resource_ready(InstId, 5000) of
+        ok ->
+            stop(InstId);
+        timeout ->
+            {error, timeout}
```

Comment:
Q: should the resource be stopped regardless of timeout?

Reply:
If I understood correctly, calling `stop(InstId)` here won't trigger that problem, right?  I think we could call that to avoid resource leakage, and still return the timeout error.

Something like this:

```suggestion
        ok ->
            stop(InstId);
        timeout ->
            stop(InstId),
            {error, timeout}
```

Does this make sense?